### PR TITLE
Add `validateInput()` hook to AdminUI Fields on Item Details page

### PR DESCRIPTION
**Pull request by jesstelford**
fixes https://github.com/keystonejs/keystone-5/issues/1222

`validateInput` takes similar parameters to the the server-side validation hook, with one addition `addFieldValidationWarning()`:

https://github.com/keystonejs/keystone-5/blob/admin-ui-client-hooks/packages/fields/src/Controller.js#L24-L57

The different between warnings and errors:

- Errors disable saving altogether.
- Warnings disable saving _on the first try_. Subsequent tries (without any edits) will save.
- If the user edits any fields after an error or warning is raised, `validateInput()` will be run again and the cycle start over.

**Errors**

admin-ui-errors

**Warnings**

admin-ui-warnings

### CODE REVIEW

File: packages/fields/src/types/Password/views/Field.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /** @jsx jsx */
 
 import { jsx } from '@emotion/core';
-import { Component, createRef } from 'react';
+import { Component, createRef, Fragment } from 'react';
 
 import { FieldContainer, FieldLabel, FieldInput } from '@arch-ui/fields';
```
Comment (jesstelford) [inline packages/fields/src/types/Password/views/Field.js]:
```suggestion
import { FieldContainer, FieldLabel, FieldInput } from '@arch-ui/fields';
import { Alert } from '@arch-ui/alert';
```

File: packages/fields/src/types/Password/views/Field.js [hunk 2]
```diff
@@ -39,13 +44,39 @@ export default class PasswordField extends Component {
...
     return (
       <FieldContainer>
         <FieldLabel htmlFor={htmlID}>{field.label}</FieldLabel>
+        {errors.length ? (
+          <Fragment>
+            <div>
+              {errors.length} Error{errors.length !== 1 ? 's' : ''}:
+            </div>
+            {errors.map(({ message, data }) => (
+              <div key={message}>
```
Comment (jesstelford) [inline packages/fields/src/types/Password/views/Field.js]:
```suggestion
              <Alert appearance="danger" key={message}>
```

File: packages/fields/src/types/Password/views/Field.js [hunk 3]
```diff
@@ -39,13 +44,39 @@ export default class PasswordField extends Component {
...
+                {message} - {JSON.stringify(data)}
+              </div>
```
Comment (jesstelford) [inline packages/fields/src/types/Password/views/Field.js]:
```suggestion
              </Alert>
```

File: packages/fields/src/types/Password/views/Field.js [hunk 4]
```diff
@@ -39,13 +44,39 @@ export default class PasswordField extends Component {
...
+          </Fragment>
+        ) : null}
+
+        {warnings.length ? (
+          <Fragment>
+            <div>
+              {warnings.length} Warning{warnings.length !== 1 ? 's' : ''}:
+            </div>
+            {warnings.map(({ message, data }) => (
+              <div key={message}>
```
Comment (jesstelford) [inline packages/fields/src/types/Password/views/Field.js]:
```suggestion
              <Alert appearance="warning" key={message}>
```

File: packages/fields/src/types/Password/views/Field.js [hunk 5]
```diff
@@ -39,13 +44,39 @@ export default class PasswordField extends Component {
...
+                {message} - {JSON.stringify(data)}
+              </div>
```
Comment (jesstelford) [inline packages/fields/src/types/Password/views/Field.js]:
```suggestion
              </Alert>
```

File: packages/app-admin-ui/client/pages/Item/index.js [hunk 6]
```diff
@@ -136,6 +147,65 @@ const ItemDetails = withRouter(
...
+      // have changed since last time we checked.
+      if (!countArrays(validationWarnings)) {
+        const errors = {};
+        const warnings = {};
+
+        let totalErrors = 0;
+        let totalWarnings = 0;
+
+        // Only validate fields whos values have changed
+        Object.keys(data).forEach(path => {
```
[the remark this answers is not in the document]
Comment (dcousens) [reply packages/app-admin-ui/client/pages/Item/index.js]: `data` is a poor name,  maybe `changedValues` ?

File: packages/fields/src/types/Password/views/Controller.js [hunk 7]
```diff
@@ -21,4 +21,22 @@ export default class PasswordController extends FieldController {
...
   ];
+
+  serialize = data => {
+    // discard the "confirm" since we only need one version of the password
+    return data[this.path] ? data[this.path].inputPassword : undefined;
+  };
+
+  validateInput = ({ originalInput, addFieldValidationError, addFieldValidationWarning }) => {
+    if (originalInput[this.path].inputPassword !== originalInput[this.path].inputConfirm) {
+      addFieldValidationError('Passwords do not match');
```
Comment (dcousens) [inline packages/fields/src/types/Password/views/Controller.js]: `return`?

**Comment by SeanDoyleGit** [comment]
@jesstelford this PR is ready 😄

**Comment by VinayaSathyanarayana** [comment]
We should consider https://github.com/keystonejs/keystone/issues/4787

**Comment by jesstelford** [comment]
@VinayaSathyanarayana that's a great suggestion! Can you open an issue on this repo so we don't lose track of the idea? Thanks 👍

**Comment by JedWatson** [comment]
@jesstelford we should have 4 styles for the flags - `success`, `info`, `warning` and `error`

Based on a visual review of the gifs you included (awesome you added them 👏) I think it should be clearer when there's an error that blocks submitting the form vs. a warning that you can override.

Let's use the red `error` style for errors, and the orange `warning` style for warnings

(If the videos are out of date and you subsequently did this, disregard, I'm just starting the code review now)

File: packages/app-admin-ui/client/pages/Item/index.js [hunk 8]
```diff
@@ -136,6 +155,45 @@ const ItemDetails = withRouter(
...
+          if (totalErrors > 0) {
+            messages.push(`${totalErrors} error${totalErrors > 1 ? 's' : ''}`);
+          }
+          if (totalWarnings > 0) {
+            messages.push(`${totalWarnings} warning${totalWarnings > 1 ? 's' : ''}`);
+          }
+
+          toastManager.add(`Validation failed: ${messages.join(' and ')}.`, {
+            autoDismiss: true,
+            appearance: errors.length ? 'error' : 'warning',
```
[the remark this answers is not in the document]
Comment (jesstelford) [reply packages/app-admin-ui/client/pages/Item/index.js]: Yeah, the video was an early WIP version. Looks much nicer now with the toast + `<Alert>` component 👍

**Pull request merged**